### PR TITLE
[#71] Implement custom accessors

### LINKED ISSUE #71: Feature: ability to add custom accessors
As a developer, I would like to define custom accessors, so that I can work with env vars in formats specific to my application's requirements.

For example:

```js
const env = from(process.env, {
  asNodes: (raiseError, value, type = Set) => {
    return new type(
      value.split(',')
        .map(s => s.trim())
        .map(s => s.split(':', 2))
        .map(([ip, port = 80]) => ({ ip, port }))
    )
  }
})

const bootnodes = env.get('BOOTNODES').asNodes()
```

Proposed acceptance criteria:
- A developer can call `from()` and provide an optional object as a second parameter.  The object defines additional accessors for any value returned by that instance's `get()`.
- If the name matches an existing accessor, the existing accessor is replaced by the new one for all subsequent calls to `env.get()`.

### PR DESCRIPTION
**Pull request by todofixthis**
**Note:** depends on #75 

closes #71

* Adds an optional `extraAccessors` param to `from()` that allows attaching a custom accessor to that instance.  Any variables subsequently gotten from that instance will have those custom accessors available.
  * Each instance has its own set of custom accessors.
  * It is allowed to override custom or even built-in accessors.

Example:
```
± % BOOTNODES="127.0.0.1, 192.168.1.1:8000" node
> const { from } = require('.')

> const env = from(process.env, {
... asNodes: (raiseError, value, type = Set) => {
.....   return new type(
.......     value.split(',')
.......       .map(s => s.trim())
.......       .map(s => s.split(':', 2))
.......       .map(([ip, port = 80]) => ({ ip, port }))
.......   )
..... }
... })

> const bootnodes = env.get('BOOTNODES').asNodes()

> bootnodes
Set {
  { ip: '127.0.0.1', port: 80 },
  { ip: '192.168.1.1', port: '8000' } }
```

See unit tests for additional examples.

### CODE REVIEW

**Comment by todofixthis** [comment]
Review requested:
- I'm completely new to TypeScript, however, so I wasn't sure what (if any) changes to make to `env-var.d.ts` and `test/index.ts`.
- Code coverage check fails, but I'm not familiar enough with the toolset to interpret the output.

File: lib/variable.js [hunk 1]
```diff
@@ -8,7 +8,7 @@ const base64Regex = /^([A-Za-z0-9+/]{4})*([A-Za-z0-9+/]{4}|[A-Za-z0-9+/]{3}=|[A-
  * the variable you wish to have returned
  * @return {Object}
  */
-module.exports = function getVariableAccessors (container, varName, defValue) {
+module.exports = function getVariableAccessors (container, varName, defValue, extraAccessors = {}) {
```
Comment (evanshortiss) [inline lib/variable.js]: @todofixthis I think removing the default param here will fix the other build issue with code coverage.

File: env-var.js [hunk 2]
```diff
@@ -9,15 +9,27 @@ const variable = require('./lib/variable')
  * @return {Object} a new module instance
  */
 const from = (container) => {
+  var _extraAccessors = {}
```
Comment (evanshortiss) [inline env-var.js]: Can you make this `const`?

**Comment by evanshortiss** [comment]
@todofixthis the PR fails on Node.js 6 since `Object.entries` isn't available. This ok though since 6 is no longer LTS. Can you delete [these lines](https://github.com/evanshortiss/env-var/blob/master/.travis.yml#L8-L9) from `travis.yml` in your PR?

The other issue is code coverage dropped

**Comment by evanshortiss** [comment]
Don't worry about the TS stuff, I'm happy to look at it if you're not familiar 😃

**Comment by todofixthis** [comment]
Ah, I should also add requisite documentation to the README while I'm at it 😅 

Will tackle these items tomorrow 😺

**Comment by evanshortiss** [comment]
Oh right, I forgot README. Blaming that on my jet lag...

Thanks!

**Comment by todofixthis** [comment]
@evanshortiss All changes have been applied to this PR.  Just need some help with the TypeScript stuffs, and then this is ready for final review 😺

**Comment by evanshortiss** [comment]
@todofixthis I've been trying to work out some TS types for this, but it's a tricky one since it mutates the underlying env instance. Mutations usually aren't TypeScript friendly, and it's being super unfriendly here. Obviously not ideal news, given your great work so far. What do you think of the changes described below? 😐

A solution I could get working with TypeScript looks similar to this in JS, since it returns a new env instance instead of mutating one. It also needs to directly throw `EnvVarError` instead of `raiseError`:

```js
const { from, EnvVarError } = require('env-var')

const env = from(process.env, {
  asEmailComponents: (value) => {
	const parts = value.split('@')

	if (parts.length != 2) {
	  throw new EnvVarError('probably not an email')
	} else {
	  return {
        username: parts[0],
        domain: parts[1]
      }
    }
  }
})

const emailParts = env.get('EMAIL').asEmailComponents()
```

We don't have to use `from`. Maybe renaming it `create` makes sense, but I think this gets the idea across. 

Here's what it looks like in TypeScript:

```ts
import { from, AccessorFn, EnvVarError } from 'env-var'

interface EmailParts {
  username: string
  domain: string
}

// EmailParts is the desired return structure
// The empty object 
const asEmailComponents: AccessorFn<EmailParts> = (value) => {
  const parts = value.split('@')

  if (parts.length != 2) {
    throw new EnvVarError('maybe not an email')
  } else {
    return {
      username: parts[0],
      domain: parts[1]
    }
  }
}

const env = from(process.env, {
  asEmailComponents
})

const emailParts = env.get('EMAIL').asEmailComponents()
```

This complicates things a bit implementation-wise since `generateAccessors` needs to pass `raiseError` to built-in accessors, but not pass it to the added accessors.

**Comment by todofixthis** [comment]
Interesting.  Actually, I was going to ask about `raiseError` — wondering if it might be possible to have `generateAccessor()` catch exceptions raised by the accessor and handle them accordingly.

I'll see if I can whip up an implementation for that, which might make this one easier to merge.

**Comment by todofixthis** [comment]
Created https://github.com/evanshortiss/env-var/pull/75 to replace `raiseError` with explicitly throwing errors.  Let me know if that looks good, and if so I'll merge that into this one.

**Comment by todofixthis** [comment]
I will also modify the code so that adding extra accessors happens at initialisation time 😺

**Comment by evanshortiss** [comment]
@todofixthis sounds good to me. That other PR is good to go, I just have one question on it before merging 👍

**Comment by todofixthis** [comment]
Woohoo!  I'll make the requested change tomorrow 😺

**Comment by todofixthis** [comment]
#75 is good to go 😺   I've gone ahead and merged that branch into this one as well, to make sure that we can merge both of them into `master` cleanly.

**Comment by evanshortiss** [comment]
@todofixthis awesome work, and thanks for your patience with the TS compatibility bits 👍

I'll merge this, sort out the types, and release as 5.1.0 ASAP.

**Pull request merged**